content/news/2021-04-04-bevy-0.5/index.md

```diff
@@ -0,0 +1,1254 @@
+
+_But_ this merge did create problems for people directly interacting with `World`. What if you need mutable access to multiple resources at the same time? `world.get_resource_mut()` borrows World mutably, which prevents multiple mutable accesses! We solved this with `WorldCell`. 
+
+### WorldCell
+
+WorldCell applies the "access control" concept used by Systems to direct world access:
+
+```rust
+let world_cell = world.cell();
+let a = world_cell.get_resource_mut::<i32>().unwrap();
+let b = world_cell.get_resource_mut::<f64>().unwrap();
+```
+
+This adds cheap runtime checks (a sparse set lookup of `ArchetypeComponentId` with a counter to indicate the number of active borrows) to ensure that world accesses do not conflict with each other. Each operation returns a `WorldBorrow<'w, T>` or `WorldBorrowMut<'w, T>` wrapper type, which will release the relevant ArchetypeComponentId resources when dropped.
+
+WorldCell does _not_ use atomic operations. It is non-send, does a mutable borrow of World to prevent other accesses, and uses a simple `Rc<RefCell<ArchetypeComponentAccess>>` wrapper in each WorldBorrow pointer. 
+
+We made this a separate api to enable users to decide what tradeoffs they want. Direct World access has stricter lifetimes, but it is more efficient and does compile time access control. `WorldCell` has looser lifetimes, but incurs a _small_ runtime penalty as a result. 
+
+The api is currently limited to resource access, but it will be extended to queries / entity component access in the future.
+
+### Resource Scopes
+
+WorldCell does not yet support component queries, and even when it does there will sometimes be legitimate reasons to want a mutable world ref _and_ a mutable resource ref (ex: bevy_render and bevy_scene both need this). In these cases we could always drop down to the unsafe `world.get_resource_unchecked_mut()`, but that is not ideal!
+
+Instead developers can use a "resource scope"
+
+```rust
+world.resource_scope(|world: &mut World, mut a: Mut<A>| {
+})
+```
+
+This temporarily removes the `A` resource from `World`, provides mutable pointers to both, and re-adds A to World when finished. Thanks to the move to ComponentIds/sparse sets, this is a cheap operation.
+
+If multiple resources are required, scopes can be nested. We could also consider adding a "resource tuple" to the api if this pattern becomes common and the boilerplate gets nasty.
+
+### Query Conflicts Use ComponentId Instead of ArchetypeComponentId
+
+For safety reasons, systems cannot contain queries that conflict with each other without wrapping them in a QuerySet. In **Bevy 0.4**, we used ArchetypeComponentIds to determine conflicts. This was nice because it could take into account filters:
```

Comment:
```suggestion
For safety reasons, systems cannot contain queries that conflict with each other without wrapping them in a `QuerySet`. In **Bevy 0.4**, we used `ArchetypeComponentIds` to determine conflicts. This was nice because it could take into account filters:
```